### ISSUE TITLE
[Feature] Pseudocode Proper Genetics Algorithm

### ISSUE THREAD

**Issue by weltvagabund** [zininzinin/AutoTrimps#12]
Here the dump from reddit. If this is implemented right, you could also fix #11 

```
[Options] // To be set in Automation Tab
Desired Breedtime [Box in s] // Default 5s
Try to reach Breedtime before Genetics (World 70) is available [Y/N Checkbox] // Default Y
 Fire to much Geneticists [Y/N Checkbox] // Default N
 Autobuy Nurseries [-1/0/Limit] // If a Limit is Set here buy until Limit, if 0 is set dont buy at all.

 Do // Loop Speed 1000 ms or slower. No need to run this super-fast. The whole system needs time to stabilize after change
 [Check Breed Time Routine] //Maybe there is a ingame variable for that nowadays?
 [Check if first Potency Update is unlocked or Bought] //Before World 5 there is no way to influence the whole thing, and Trimp Numbers and everything else is changing fast anyhow
 //Cases
 If Breedtime is under Desired_Breedtime
     Dont Buy Potency Upgrade
     Dont Buy Nurseries if available
     Buy Geneticists if available // I would only buy one per Iteration of the Loop. This is not as Time Critical as other things.
     //If no Workers are available, I would fire a Farmer, not a Scientist.
Next

If Breedtime is over Desired_Breedtime
    Buy Potency Upgrade // 1 per Iteration of Loop (1st Prio)
    Buy Nursery // 1 per Iteration of Loop (2nd Prio, only if there are no more Potency)
    Fire Geneticist if Option is Y // 1 per Loop (if Option is N just wait until enough Nurseries/Potency Upgrades are available) (3rd Prio)
Next
End
```

**Comment by spindrjr** [comment]
In the interest of trying to keep the number of settings in this script under control, do we really think having a option to disable firing of geneticists is necessary? If we put in a check to see if housing is currently being built, I think we can minimize any losses from auto-firing to the point that they are inconsequential.

In other thoughts on this, it definitely needs nursery logic, and I would be inclined to disable building of nurseries if the user has any levels of the anticipation perk until geneticists are available.

On the subject of anticipation, in lieu of programming logic in to delay purchasing of potency, I'd like to see if we can't make a suicide function to suicide armies and artificially reach 30 anticipation stacks prior to geneticists.

And lastly I think this function should probably automatically override any breed timer higher than about 5 seconds if Electricity (nom?) is active (as well as override the no-nursery purchasing before geneticists).

**Comment by weltvagabund** [comment]
In Nom it is Critical to do at least 5% Enemy HP Damage per Death to Progress. There those Extra Percent from Anticipation are critical. If you want to go this Road, you could also calculate the Optimal Genetics for each Cell as a Function of ownDamage, ownBlock, Stance, Anticipation, Breed Time, enemyDamage and enemyHP. If this is done properly you could combine it with Stances, but that gets fricking complicated....

**Comment by spindrjr** [comment]
I pushed an update with an enhanced genetics code. The suicide function to artificially reach max anticipation stacks would end up being far too 'cheaty' for my tastes so I went with your suggestion to hold off potency upgrades/nurseries. 

As programmed, it is a single option to control breed timer prior to geneticists, though the option has no UI access as of now and is hardcoded on. No option to disable firing of geneticists as I see no reason for it (though I'm open to persuasion still). You can find some more details in the commit notes. Give it a try and let me know if you find any issues.

Closing this issue.

**Comment by weltvagabund** [comment]
Hej, I looked only at the code right now. It looks really exciting. I am still full of hope, that the new UI will be done soon, so I am holding of the testing right now. But maybe i give it a run this afternoon.

But I noticed one thing with the Auto Buy Nurseries, that is still present in the current version. High Level Nurseries consume a LOT of Gems. So they significantly delay the buying of new Warp/Gigastations, if left run to the maximum. (buy one Warp, buy Nursery, buy Nursery, buy Nursery, buy Warp, ...) With waiting times of 5 to 15 min, this really puts a dent in Coordiantion Updates. Maybe we should insert a Cost Break in somewhere, as of use max 25% of Gem Production, or if Gem Storage > Highest Warpstation ever bought *2.

I am not sure how to define the Breakpoint now, but for the sake of fast progress i always prefer Trimps over Nurseries.
